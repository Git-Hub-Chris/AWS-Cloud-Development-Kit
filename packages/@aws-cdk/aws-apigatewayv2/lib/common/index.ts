export * from './api-mapping';
export * from './authorizer';
export * from './domain-name';
export * from './integration';
export * from './json-schema';
export * from './route';
export * from './stage';