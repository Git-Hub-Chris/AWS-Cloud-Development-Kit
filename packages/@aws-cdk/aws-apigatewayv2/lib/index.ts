// AWS::APIGatewayv2 CloudFormation Resources:
export * from './apigatewayv2.generated';
export * from './api';
export * from './api-mapping';
export * from './authorizer';
export * from './deployment';
export * from './domain-name';
export * from './http-api';
export * from './integration';
export * from './integration-response';
export * from './integrations';
export * from './json-schema';
export * from './model';
export * from './route';
export * from './route-response';
export * from './stage';
export * from './web-socket-api';
