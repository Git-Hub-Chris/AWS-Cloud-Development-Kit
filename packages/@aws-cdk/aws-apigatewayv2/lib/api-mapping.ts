import { Construct, IResource, Resource } from '@aws-cdk/core';

import { IApi } from './api';
import { CfnApiMapping } from './apigatewayv2.generated';
import { IDomainName } from './domain-name';
import { IStage } from './stage';

/**
 * Defines the contract for an Api Gateway V2 Api Mapping.
 */
export interface IApiMapping extends IResource {
  /**
   * The ID of this API Gateway Api Mapping.
   * @attribute
   */
  readonly apiMappingId: string;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Api Mapping.
 *
 * This interface is used by the helper methods in `Api` and the sub-classes
 */
export interface ApiMappingOptions {
  /**
   * The API mapping key.
   *
   * @default - no API mapping key
   */
  readonly apiMappingKey?: string;

  /**
   * The associated domain name
   */
  readonly domainName: IDomainName |Â string;

  /**
   * The API stage.
   */
  readonly stage: IStage;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Api Mapping.
 */
export interface ApiMappingProps extends ApiMappingOptions {
  /**
   * Defines the api for this deployment.
   */
  readonly api: IApi;
}

/**
 * An Api Mapping for an API. An API mapping relates a path of your custom domain name to a stage of your API.
 *
 * A custom domain name can have multiple API mappings, but the paths can't overlap.
 *
 * A custom domain can map only to APIs of the same protocol type.
 */
export class ApiMapping extends Resource implements IApiMapping {

  /**
   * Creates a new imported API
   *
   * @param scope scope of this imported resource
   * @param id identifier of the resource
   * @param apiMappingId Identifier of the API Mapping
   */
  public static fromApiMappingId(scope: Construct, id: string, apiMappingId: string): IApiMapping {
    class Import extends Resource implements IApiMapping {
      public readonly apiMappingId = apiMappingId;
    }

    return new Import(scope, id);
  }

  /**
   * The ID of this API Gateway Api Mapping.
   */
  public readonly apiMappingId: string;

  protected resource: CfnApiMapping;

  constructor(scope: Construct, id: string, props: ApiMappingProps) {
    super(scope, id);

    this.resource = new CfnApiMapping(this, 'Resource', {
      apiId: props.api.apiId,
      domainName: ((typeof(props.domainName) === 'string') ? props.domainName : props.domainName.domainName),
      stage: props.stage.stageName,
      apiMappingKey: props.apiMappingKey,
    });
    this.apiMappingId = this.resource.ref;
  }
}