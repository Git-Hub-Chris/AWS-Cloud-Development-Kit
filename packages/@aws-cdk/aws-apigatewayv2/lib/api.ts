import { Construct, IResource, Resource, Stack } from '@aws-cdk/core';

import { CfnApi } from './apigatewayv2.generated';
import { Deployment } from './deployment';
import { Integration } from './integration';
import { LambdaIntegration, LambdaIntegrationOptions } from './integrations/lambda-integration';
import { JsonSchema } from './json-schema';
import { Model, ModelOptions } from './model';
import { IRoute, KnownRouteSelectionExpression } from './route';
import { IStage, Stage, StageOptions } from './stage';

/**
 * Available protocols for ApiGateway V2 APIs (currently only 'WEBSOCKET' is supported)
 */
export enum ProtocolType {
  /**
   * WebSocket API
   */
  WEBSOCKET = "WEBSOCKET",

  /**
   * HTTP API
   */
  HTTP = "HTTP"
}

/**
 * Specifies how to interpret the base path of the API during import
 */
export enum BasePath {
  /**
   * Ignores the base path
   */
  IGNORE = "ignore",

  /**
   * Prepends the base path to the API path
   */
  PREPEND = "prepend",

  /**
   * Splits the base path from the API path
   */
  SPLIT = "split"
}

/**
 * This expression is evaluated when the service determines the given request should proceed
 * only if the client provides a valid API key.
 */
export enum KnownApiKeySelectionExpression {
  /**
   * Uses the `x-api-key` header to get the API Key
   */
  HEADER_X_API_KEY = "$request.header.x-api-key",

  /**
   * Uses the `usageIdentifier` property of the current authorizer to get the API Key
   */
  AUTHORIZER_USAGE_IDENTIFIER = " $context.authorizer.usageIdentifierKey"
}

/**
 * The BodyS3Location property specifies an S3 location from which to import an OpenAPI definition.
 */
export interface BodyS3Location {
  /**
   * The S3 bucket that contains the OpenAPI definition to import.
   */
  readonly bucket: string;

  /**
   * The Etag of the S3 object.
   *
   * @default - no etag verification
   */
  readonly etag?: string;

  /**
   * The key of the S3 object.
   */
  readonly key: string;

  /**
   * The version of the S3 object.
   *
   * @default - get latest version
   */
  readonly version?: string;
}

/**
 * The CorsConfiguration property specifies a CORS configuration for an API.
 */
export interface CorsConfiguration {
  /**
   * Specifies whether credentials are included in the CORS request.
   *
   * @default false
   */
  readonly allowCredentials?: boolean;

  /**
   * Represents a collection of allowed headers.
   *
   * @default - no headers allowed
   */
  readonly allowHeaders?: string[];

  /**
   * Represents a collection of allowed HTTP methods.
   *
   * @default - no allowed methods
   */
  readonly allowMethods: string[];

  /**
   * Represents a collection of allowed origins.
   *
   * @default - get allowed origins
   */
  readonly allowOrigins?: string[];

  /**
   * Represents a collection of exposed headers.
   *
   * @default - get allowed origins
   */
  readonly exposeHeaders?: string[];

  /**
   * The number of seconds that the browser should cache preflight request results.
   *
   * @default - get allowed origins
   */
  readonly maxAge?: number;
}

/**
 * Defines the contract for an Api Gateway V2 Api.
 */
export interface IApi extends IResource {
  /**
   * The ID of this API Gateway Api.
   * @attribute
   */
  readonly apiId: string;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Api.
 */
export interface ApiProps {
  /**
   * Indicates if a Deployment should be automatically created for this API,
   * and recreated when the API model (route, integration) changes.
   *
   * Since API Gateway deployments are immutable, When this option is enabled
   * (by default), an AWS::ApiGatewayV2::Deployment resource will automatically
   * created with a logical ID that hashes the API model (methods, resources
   * and options). This means that when the model changes, the logical ID of
   * this CloudFormation resource will change, and a new deployment will be
   * created.
   *
   * If this is set, `latestDeployment` will refer to the `Deployment` object
   * and `deploymentStage` will refer to a `Stage` that points to this
   * deployment. To customize the stage options, use the `deployOptions`
   * property.
   *
   * A CloudFormation Output will also be defined with the root URL endpoint
   * of this REST API.
   *
   * @default true
   */
  readonly deploy?: boolean;

  /**
   * Options for the API Gateway stage that will always point to the latest
   * deployment when `deploy` is enabled. If `deploy` is disabled,
   * this value cannot be set.
   *
   * @default - default options
   */
  readonly deployOptions?: StageOptions;

  /**
   * Retains old deployment resources when the API changes. This allows
   * manually reverting stages to point to old deployments via the AWS
   * Console.
   *
   * @default false
   */
  readonly retainDeployments?: boolean;

  /**
   * A name for the API Gateway Api resource.
   *
   * @default - ID of the Api construct.
   */
  readonly apiName?: string;

  /**
   * Specifies how to interpret the base path of the API during import.
   *
   * Supported only for HTTP APIs.
   * @default 'ignore'
   */
  readonly basePath?: BasePath;

  /**
   * The OpenAPI definition.
   *
   * To import an HTTP API, you must specify `body` or `bodyS3Location`.
   *
   * Supported only for HTTP APIs.
   * @default - `bodyS3Location` if defined, or no import
   */
  readonly body?: string;

  /**
   * The S3 location of an OpenAPI definition.
   *
   * To import an HTTP API, you must specify `body` or `bodyS3Location`.
   *
   * Supported only for HTTP APIs.
   * @default - `body` if defined, or no import
   */
  readonly bodyS3Location?: BodyS3Location;

  /**
   * The S3 location of an OpenAPI definition.
   *
   * To import an HTTP API, you must specify `body` or `bodyS3Location`.
   *
   * Supported only for HTTP APIs.
   * @default - `body` if defined, or no import
   */
  readonly corsConfiguration?: CorsConfiguration;

  /**
   * Available protocols for ApiGateway V2 APIs
   *
   * @default - required unless you specify an OpenAPI definition
   */
  readonly protocolType?: ProtocolType | string;

  /**
   * Expression used to select the route for this API
   *
   * @default - '${request.method} ${request.path}' for HTTP APIs, required for Web Socket APIs
   */
  readonly routeSelectionExpression?: KnownRouteSelectionExpression | string;

  /**
   * Expression used to select the Api Key to use for metering
   *
   * Supported only for WebSocket APIs
   *
   * @default - No Api Key
   */
  readonly apiKeySelectionExpression?: KnownApiKeySelectionExpression |Â string;

  /**
   * A description of the purpose of this API Gateway Api resource.
   *
   * @default - No description.
   */
  readonly description?: string;

  /**
   * Indicates whether schema validation will be disabled for this Api
   *
   * @default false
   */
  readonly disableSchemaValidation?: boolean;

  /**
   * Indicates the version number for this Api
   *
   * @default false
   */
  readonly version?: string;

  /**
   * Specifies whether to rollback the API creation (`true`) or not (`false`) when a warning is encountered
   *
   * @default false
   */
  readonly failOnWarnings?: boolean;
}

/**
 * Represents a WebSocket API in Amazon API Gateway v2.
 *
 * Use `addModel` and `addLambdaIntegration` to configure the API model.
 *
 * By default, the API will automatically be deployed and accessible from a
 * public endpoint.
 */
export class Api extends Resource implements IApi {

  /**
   * Creates a new imported API
   *
   * @param scope scope of this imported resource
   * @param id identifier of the resource
   * @param apiId Identifier of the API
   */
  public static fromApiId(scope: Construct, id: string, apiId: string): IApi {
    class Import extends Resource implements IApi {
      public readonly apiId = apiId;
    }

    return new Import(scope, id);
  }

  /**
   * The ID of this API Gateway Api.
   */
  public readonly apiId: string;

  /**
   * API Gateway stage that points to the latest deployment (if defined).
   */
  public deploymentStage?: Stage;

  protected resource: CfnApi;
  protected deployment?: Deployment;

  constructor(scope: Construct, id: string, props?: ApiProps) {
    if (props === undefined) {
      props = {};
    }
    super(scope, id, {
      physicalName: props.apiName || id,
    });

    if (props.protocolType === undefined && props.body === null && props.bodyS3Location === null) {
      throw new Error("You must specify a protocol type, or import an Open API definition (directly or from S3)");
    }

    switch (props.protocolType) {
      case ProtocolType.WEBSOCKET: {
        if (props.basePath !== undefined) {
          throw new Error('"basePath" is only supported with HTTP APIs');
        }
        if (props.body !== undefined) {
          throw new Error('"body" is only supported with HTTP APIs');
        }
        if (props.bodyS3Location !== undefined) {
          throw new Error('"bodyS3Location" is only supported with HTTP APIs');
        }
        if (props.corsConfiguration !== undefined) {
          throw new Error('"corsConfiguration" is only supported with HTTP APIs');
        }
        if (props.routeSelectionExpression === undefined) {
          throw new Error('"routeSelectionExpression" is required for Web Socket APIs');
        }
        break;
      }
      case ProtocolType.HTTP:
      case undefined: {
        if (props.apiKeySelectionExpression !== undefined) {
          throw new Error('"apiKeySelectionExpression" is only supported with Web Socket APIs');
        }
        if (props.disableSchemaValidation !== undefined) {
          throw new Error('"disableSchemaValidation" is only supported with Web Socket APIs');
        }
        if (props.routeSelectionExpression !== undefined && props.apiKeySelectionExpression !== KnownRouteSelectionExpression.METHOD_PATH) {
          throw new Error('"routeSelectionExpression" has a single supported value for HTTP APIs: "${request.method} ${request.path}"');
        }
        break;
      }
    }

    this.resource = new CfnApi(this, 'Resource', {
      ...props,
      name: this.physicalName
    });
    this.apiId = this.resource.ref;

    const deploy = props.deploy === undefined ? true : props.deploy;
    if (deploy) {
      // encode the stage name into the construct id, so if we change the stage name, it will recreate a new stage.
      // stage name is part of the endpoint, so that makes sense.
      const stageName = (props.deployOptions && props.deployOptions.stageName) || 'prod';

      this.deployment = new Deployment(this, 'Deployment', {
        api: this,
        description: 'Automatically created by the Api construct',

        // No stageName specified, this will be defined by the stage directly, as it will reference the deployment
        retainDeployments: props.retainDeployments
      });

      this.deploymentStage = new Stage(this, `Stage.${stageName}`, {
        ...props.deployOptions,
        deployment: this.deployment,
        api: this,
        stageName,
        description: 'Automatically created by the Api construct'
      });
    } else {
      if (props.deployOptions) {
        throw new Error(`Cannot set 'deployOptions' if 'deploy' is disabled`);
      }
    }
  }

  /**
   * API Gateway deployment that represents the latest changes of the API.
   * This resource will be automatically updated every time the REST API model changes.
   * This will be undefined if `deploy` is false.
   */
  public get latestDeployment() {
    return this.deployment;
  }

  /**
   * Defines an API Gateway Lambda integration.
   * @param id The construct id
   * @param props Lambda integration options
   */
  public addLambdaIntegration(id: string, props: LambdaIntegrationOptions): Integration {
    return new LambdaIntegration(this, id, { ...props, api: this });
  }

  /**
   * Defines a model for this Api Gateway.
   * @param schema The model schema
   * @param props The model integration options
   */
  public addModel(schema: JsonSchema, props?: ModelOptions): Model {
    return new Model(this, `Model.${schema.title}`, {
      ...props,
      modelName: schema.title,
      api: this,
      schema
    });
  }

  /**
   * Returns the ARN for a specific route and stage.
   *
   * @param route The route for this ARN ('*' if not defined)
   * @param stage The stage for this ARN (if not defined, defaults to the deployment stage if defined, or to '*')
   */
  public executeApiArn(route?: IRoute, stage?: IStage) {
    const stack = Stack.of(this);
    const apiId = this.apiId;
    const routeKey = ((route === undefined) ? '*' : route.key);
    const stageName = ((stage === undefined) ?
      ((this.deploymentStage === undefined) ? '*' : this.deploymentStage.stageName) :
      stage.stageName);
    return stack.formatArn({
      service: 'execute-api',
      resource: apiId,
      sep: '/',
      resourceName: `${stageName}/${routeKey}`
    });
  }

  /**
   * Returns the ARN for a specific connection.
   *
   * @param connectionId The identifier of this connection ('*' if not defined)
   * @param stage The stage for this ARN (if not defined, defaults to the deployment stage if defined, or to '*')
   */
  public connectionsApiArn(connectionId: string = "*", stage?: IStage) {
    const stack = Stack.of(this);
    const apiId = this.apiId;
    const stageName = ((stage === undefined) ?
      ((this.deploymentStage === undefined) ? '*' : this.deploymentStage.stageName) :
      stage.stageName);
    return stack.formatArn({
      service: 'execute-api',
      resource: apiId,
      sep: '/',
      resourceName: `${stageName}/POST/${connectionId}`
    });
  }

  /**
   * Returns the client URL for this Api, for a specific stage.
   *
   * Fails if `stage` is not defined, and `deploymentStage` is not set either by `deploy` or explicitly.
   * @param stage The stage for this URL (if not defined, defaults to the deployment stage)
   */
  public clientUrl(stage?: IStage): string {
    const stack = Stack.of(this);
    let stageName: string |Â undefined;
    if (stage === undefined) {
      if (this.deploymentStage === undefined) {
        throw Error("No stage defined for this Api");
      }
      stageName = this.deploymentStage.stageName;
    } else {
      stageName = stage.stageName;
    }
    return `wss://${this.apiId}.execute-api.${stack.region}.amazonaws.com/${stageName}`;
  }

  /**
   * Returns the connections URL for this Api, for a specific stage.
   *
   * Fails if `stage` is not defined, and `deploymentStage` is not set either by `deploy` or explicitly.
   * @param stage The stage for this URL (if not defined, defaults to the deployment stage)
   */
  public connectionsUrl(stage?: IStage): string {
    const stack = Stack.of(this);
    let stageName: string |Â undefined;
    if (stage === undefined) {
      if (this.deploymentStage === undefined) {
        throw Error("No stage defined for this Api");
      }
      stageName = this.deploymentStage.stageName;
    } else {
      stageName = stage.stageName;
    }
    return `https://${this.apiId}.execute-api.${stack.region}.amazonaws.com/${stageName}/@connections`;
  }
}