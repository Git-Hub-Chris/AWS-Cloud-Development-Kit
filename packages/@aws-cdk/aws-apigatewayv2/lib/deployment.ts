import { CfnResource, Construct, IResource, Lazy, Resource } from '@aws-cdk/core';

import { IApi } from './api';
import { CfnDeployment } from './apigatewayv2.generated';

/**
 * Defines the contract for an Api Gateway V2 Deployment.
 */
export interface IDeployment extends IResource {
  /**
   * The ID of this API Gateway Deployment.
   * @attribute
   */
  readonly deploymentId: string;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Deployment.
 */
export interface DeploymentProps {
  /**
   * Defines the api for this deployment.
   */
  readonly api: IApi;

  /**
   * A description for this Deployment.
   *
   * @default - No description.
   */
  readonly description?: string;

  /**
   * The stage name. Stage names can only contain alphanumeric characters, hyphens, and underscores. Maximum length is 128 characters.
   *
   * @default - All stages.
   */
  readonly stageName?: string;
}

/**
 * A Deployment of an API.
 *
 * An immutable representation of an Api resource that can be called by users
 * using Stages. A deployment must be associated with a Stage for it to be
 * callable over the Internet.
 *
 * Normally, you don't need to define deployments manually. The Api
 * construct manages a Deployment resource that represents the latest model. It
 * can be accessed through `api.latestDeployment` (unless `deploy: false` is
 * set when defining the `Api`).
 *
 * If you manually define this resource, you will need to know that since
 * deployments are immutable, as long as the resource's logical ID doesn't
 * change, the deployment will represent the snapshot in time in which the
 * resource was created. This means that if you modify the RestApi model (i.e.
 * add methods or resources), these changes will not be reflected unless a new
 * deployment resource is created.
 *
 * To achieve this behavior, the method `addToLogicalId(data)` can be used to
 * augment the logical ID generated for the deployment resource such that it
 * will include arbitrary data. This is done automatically for the
 * `api.latestDeployment` deployment.
 *
 * Furthermore, since a deployment does not reference any of the API
 * resources and methods, CloudFormation will likely provision it before these
 * resources are created, which means that it will represent a "half-baked"
 * model. Use the `registerDependency(dep)` method to circumvent that. This is done
 * automatically for the `api.latestDeployment` deployment.
 */
export class Deployment extends Resource implements IDeployment {
  /**
   * Creates a new imported API Deployment
   *
   * @param scope scope of this imported resource
   * @param id identifier of the resource
   * @param deploymentId Identifier of the Deployment
   */
  public static fromDeploymentId(scope: Construct, id: string, deploymentId: string): IDeployment {
    class Import extends Resource implements IDeployment {
      public readonly deploymentId = deploymentId;
    }

    return new Import(scope, id);
  }

  /**
   * The ID of this API Gateway Deployment.
   */
  public readonly deploymentId: string;

  protected resource: CfnDeployment;

  constructor(scope: Construct, id: string, props: DeploymentProps) {
    super(scope, id);

    this.resource = new CfnDeployment(this, 'Resource', {
      apiId: props.api.apiId,
      description: props.description,
      stageName: props.stageName,
    });

    this.deploymentId = Lazy.stringValue({ produce: () => this.resource.ref });
  }

  /**
   * Adds a dependency to the Deployment node, avoiding partial deployments.
   *
   * This is done automatically for the `api.latestDeployment` deployment.
   *
   * @param dependency The construct to add in the dependency tree
   */
  public registerDependency(dependency: CfnResource) {
    this.resource.addDependsOn(dependency);
  }
}