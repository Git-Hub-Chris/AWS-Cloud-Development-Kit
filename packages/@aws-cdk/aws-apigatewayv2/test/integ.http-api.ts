import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import * as apigatewayv2 from '../lib';

// tslint:disable:max-line-length

const app = new cdk.App();

const stack = new cdk.Stack(app, 'ApiagtewayV2HttpApi');

const getbooksHandler = new lambda.Function(stack, 'MyFunc', {
  runtime: lambda.Runtime.PYTHON_3_7,
  handler: 'index.handler',
  code: new lambda.InlineCode(`
import json
def handler(event, context):
      return {
        'statusCode': 200,
        'body': json.dumps(event)
      }`),
});

const getbookReviewsHandler = new lambda.Function(stack, 'RootFunc', {
  runtime: lambda.Runtime.PYTHON_3_7,
  handler: 'index.handler',
  code: new lambda.InlineCode(`
import json, os
def handler(event, context):
      whoami = os.environ['WHOAMI']
      http_path = os.environ['HTTP_PATH']
      return {
        'statusCode': 200,
        'body': json.dumps({ 'whoami': whoami, 'http_path': http_path })
      }`),
  environment: {
    WHOAMI: 'root',
    HTTP_PATH: '/',
  },
});

const rootUrl = 'https://checkip.amazonaws.com';
const defaultUrl = 'https://aws.amazon.com';

// create a basic HTTP API with http proxy integration as the $default route
const api = new apigatewayv2.HttpApi(stack, 'HttpApi', {
  defaultTarget: { uri: defaultUrl },
});

api.addRoutes(['GET /', 'POST /'], api.addHttpIntegration('RootInteg', { url: rootUrl }));
api.addRoute({ method: apigatewayv2.HttpMethod.GET, path: '/books'Â }, api.addLambdaIntegration('getbooksInteg', { handler : getbooksHandler }));
api.addRoute('GET /books/reviews', api.addLambdaIntegration('getBookReviewInteg', { handler: getbookReviewsHandler }));
