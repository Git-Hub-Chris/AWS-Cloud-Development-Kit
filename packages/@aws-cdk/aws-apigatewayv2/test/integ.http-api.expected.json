{
  "Resources": {
    "MyFuncServiceRole54065130": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "MyFunc8A243A2C": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "\nimport json\ndef handler(event, context):\n      return {\n        'statusCode': 200,\n        'body': json.dumps(event)\n      }"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "MyFuncServiceRole54065130",
            "Arn"
          ]
        },
        "Runtime": "python3.7"
      },
      "DependsOn": [
        "MyFuncServiceRole54065130"
      ]
    },
    "MyFuncApiPermissionApiagtewayV2HttpApigetbooksInteglambdaintegration4555698DC6292E59": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MyFunc8A243A2C",
            "Arn"
          ]
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region"
              },
              ":",
              {
                "Ref": "AWS::AccountId"
              },
              ":",
              {
                "Ref": "HttpApiF5A9A8A7"
              },
              "/",
              {
                "Ref": "HttpApiDefaultStage3EEB07D6"
              },
              "/*"
            ]
          ]
        }
      }
    },
    "RootFuncServiceRoleE4AA9E41": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "RootFuncF39FB174": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "\nimport json, os\ndef handler(event, context):\n      whoami = os.environ['WHOAMI']\n      http_path = os.environ['HTTP_PATH']\n      return {\n        'statusCode': 200,\n        'body': json.dumps({ 'whoami': whoami, 'http_path': http_path })\n      }"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "RootFuncServiceRoleE4AA9E41",
            "Arn"
          ]
        },
        "Runtime": "python3.7",
        "Environment": {
          "Variables": {
            "WHOAMI": "root",
            "HTTP_PATH": "/"
          }
        }
      },
      "DependsOn": [
        "RootFuncServiceRoleE4AA9E41"
      ]
    },
    "RootFuncApiPermissionApiagtewayV2HttpApigetBookReviewInteglambdaintegration0ABD780C9E2600E7": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "RootFuncF39FB174",
            "Arn"
          ]
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region"
              },
              ":",
              {
                "Ref": "AWS::AccountId"
              },
              ":",
              {
                "Ref": "HttpApiF5A9A8A7"
              },
              "/",
              {
                "Ref": "HttpApiDefaultStage3EEB07D6"
              },
              "/*"
            ]
          ]
        }
      }
    },
    "HttpApiF5A9A8A7": {
      "Type": "AWS::ApiGatewayV2::Api",
      "Properties": {
        "Name": "HttpApi",
        "ProtocolType": "HTTP",
        "RouteSelectionExpression": "${request.method} ${request.path}"
      }
    },
    "HttpApiDeployment2ABEB12F": {
      "Type": "AWS::ApiGatewayV2::Deployment",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "Description": "Automatically created by the Api construct"
      },
      "DependsOn": [
        "HttpApidefaultroute8BBAF64E",
        "HttpApidefaulthttpintegration5F125B3A",
        "HttpApiGETrouteF5E07819",
        "HttpApiGETbooksreviewsrouteCAD382E0",
        "HttpApiGETbooksroute2D32C4F3",
        "HttpApigetBookReviewInteglambdaintegrationD5870D18",
        "HttpApigetbooksInteglambdaintegrationE71E55E1",
        "HttpApiPOSTroute74994293",
        "HttpApiRootInteghttpintegration15582736"
      ]
    },
    "HttpApiDefaultStage3EEB07D6": {
      "Type": "AWS::ApiGatewayV2::Stage",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "StageName": "prod",
        "DeploymentId": {
          "Ref": "HttpApiDeployment2ABEB12F"
        },
        "Description": "Automatically created by the Api construct"
      }
    },
    "HttpApidefaulthttpintegration5F125B3A": {
      "Type": "AWS::ApiGatewayV2::Integration",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "IntegrationType": "HTTP_PROXY",
        "IntegrationMethod": "ANY",
        "IntegrationUri": "https://aws.amazon.com",
        "PayloadFormatVersion": "1.0"
      }
    },
    "HttpApidefaultroute8BBAF64E": {
      "Type": "AWS::ApiGatewayV2::Route",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "RouteKey": "$default",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApidefaulthttpintegration5F125B3A"
              }
            ]
          ]
        }
      }
    },
    "HttpApiRootInteghttpintegration15582736": {
      "Type": "AWS::ApiGatewayV2::Integration",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "IntegrationType": "HTTP_PROXY",
        "IntegrationMethod": "ANY",
        "IntegrationUri": "https://checkip.amazonaws.com",
        "PayloadFormatVersion": "1.0"
      }
    },
    "HttpApiGETrouteF5E07819": {
      "Type": "AWS::ApiGatewayV2::Route",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "RouteKey": "GET /",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApiRootInteghttpintegration15582736"
              }
            ]
          ]
        }
      }
    },
    "HttpApiPOSTroute74994293": {
      "Type": "AWS::ApiGatewayV2::Route",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "RouteKey": "POST /",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApiRootInteghttpintegration15582736"
              }
            ]
          ]
        }
      }
    },
    "HttpApigetbooksInteglambdaintegrationE71E55E1": {
      "Type": "AWS::ApiGatewayV2::Integration",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationMethod": "POST",
        "IntegrationUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region"
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "MyFunc8A243A2C",
                  "Arn"
                ]
              },
              "/invocations"
            ]
          ]
        },
        "PayloadFormatVersion": "1.0"
      }
    },
    "HttpApiGETbooksroute2D32C4F3": {
      "Type": "AWS::ApiGatewayV2::Route",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "RouteKey": "GET /books",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApigetbooksInteglambdaintegrationE71E55E1"
              }
            ]
          ]
        }
      }
    },
    "HttpApigetBookReviewInteglambdaintegrationD5870D18": {
      "Type": "AWS::ApiGatewayV2::Integration",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationMethod": "POST",
        "IntegrationUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region"
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "RootFuncF39FB174",
                  "Arn"
                ]
              },
              "/invocations"
            ]
          ]
        },
        "PayloadFormatVersion": "1.0"
      }
    },
    "HttpApiGETbooksreviewsrouteCAD382E0": {
      "Type": "AWS::ApiGatewayV2::Route",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7"
        },
        "RouteKey": "GET /books/reviews",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApigetBookReviewInteglambdaintegrationD5870D18"
              }
            ]
          ]
        }
      }
    }
  }
}